version: '3.9'

services:
  traefik:
    image: traefik:v2.9
    restart: always
    command:
      - --api=true
      - --api.dashboard=true
      - --api.insecure=${TRAEFIK_API_INSECURE:-false}
      - --providers.docker=true
      - --entrypoints.http.address=:80
      - --entrypoints.https.address=:443
      - --entrypoints.http.forwardedHeaders.insecure
      - --entrypoints.https.forwardedHeaders.insecure
      - --certificatesresolvers.myresolver.acme.email=${TRAEFIK_ACME_EMAIL}
      - --certificatesresolvers.myresolver.acme.storage=/acme/acme.json
      - --certificatesresolvers.myresolver.acme.httpchallenge.entrypoint=http
      - --log.level=INFO
    ports:
      - '80:80'
      - '443:443'
      - '8080:8080'
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - traefik-acme:/acme/acme.json
    networks:
      - web

  web:
    labels:
      - traefik.enable=true
      - traefik.http.routers.web.rule=Host(`app.${APP_DOMAIN}`)
      - traefik.http.routers.web.entrypoints=http,https
      - traefik.http.routers.web.tls.certresolver=myresolver
      - traefik.http.services.web.loadbalancer.server.port=8080
    networks:
      - web
      - default

  gateway:
    labels:
      - traefik.enable=true
      - traefik.http.routers.gateway.rule=Host(`gateway.${APP_DOMAIN}`)
      - traefik.http.routers.gateway.entrypoints=http,https
      - traefik.http.routers.gateway.tls.certresolver=myresolver
      - traefik.http.services.gateway.loadbalancer.server.port=8080
    networks:
      - web
      - default

  websockets:
    labels:
      - traefik.enable=true
      - traefik.http.routers.websockets.rule=Host(`ws.${APP_DOMAIN}`)
      - traefik.http.routers.websockets.entrypoints=http,https
      - traefik.http.routers.websockets.tls.certresolver=myresolver
      - traefik.http.services.websockets.loadbalancer.server.port=4002
    networks:
      - web
      - default

networks:
  web:
    external: true

volumes:
  traefik-acme: